File: article_34/18_variadic_function.c
#include <stdio.h>
#include <stdarg.h>

void f1(const char *msg) {
    printf("%s\n", msg);
}

void f2(int count, ...) {
    int i;
    va_list args;
    va_start(args, count);

    for (i = 0; i < count; i++) {
        char *msg = va_arg(args, char *);
        printf("%s ", msg);
    }
    putchar('\n');
    va_end(args);
}

void f3(char *prefix, int count, ...) {
    int i;
    va_list args;

    printf("%s ", prefix);
    va_start(args, count);

    for (i = 0; i < count; i++) {
        char *msg = va_arg(args, char *);
        printf("%s ", msg);
    }
    putchar('\n');
    va_end(args);
}

int main(void) {
    f1("Hello");
    f2(3, "Hello", "world", "!");
    f3("Message:", 4, "Hello", "world", "again", "!");
    return 0;
}

File: article_34/18_variadic_function.go
// Seriál "Programovací jazyk Go"
//    https://www.root.cz/serialy/programovaci-jazyk-go/
//
// Třicátá čtvrtá část
//    Programovací jazyk Go pro skalní céčkaře
//    https://www.root.cz/clanky/programovaci-jazyk-go-pro-skalni-ceckare/
//
// Seznam příkladů ze třicáté čtvrté části:
//    https://github.com/tisnik/go-root/blob/master/article_34/README.md
//
// Demonstrační příklad číslo 18:
//    Variadické funkce.

package main

import "fmt"

func f1(msg string) {
	fmt.Printf("%s\n", msg)
}

func f2(parts ...string) {
	for _, val := range parts {
		fmt.Printf("%s ", val)
	}
	fmt.Println()
}

func f3(prefix string, parts ...string) {
	fmt.Printf("%s ", prefix)
	for _, val := range parts {
		fmt.Printf("%s ", val)
	}
	fmt.Println()
}

func main() {
	f1("Hello")
	f2("Hello", "world", "!")
	f3("Message:", "Hello", "world", "again", "!")
}
